File: http/http-server-annotation-processor/src/main/java/ru/tinkoff/kora/http/server/annotation/processor/extension/HttpServerRequestMapperKoraExtension.java
package ru.tinkoff.kora.http.server.annotation.processor.extension;

import com.palantir.javapoet.CodeBlock;
import com.palantir.javapoet.ParameterizedTypeName;
import com.palantir.javapoet.TypeName;
import org.jspecify.annotations.Nullable;
import ru.tinkoff.kora.http.server.annotation.processor.HttpServerClassNames;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.ExtensionResult;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.ArrayList;
import java.util.List;

public final class HttpServerRequestMapperKoraExtension implements KoraExtension {
    private final Elements elements;
    private final Types types;

    public HttpServerRequestMapperKoraExtension(Elements elements, Types types) {
        this.elements = elements;
        this.types = types;
    }

    @Nullable
    @Override
    public KoraExtensionDependencyGenerator getDependencyGenerator(RoundEnvironment roundEnvironment, TypeMirror typeMirror, String tag) {
        var typeName = TypeName.get(typeMirror);
        if (!(typeName instanceof ParameterizedTypeName ptn)) {
            return null;
        }
        if (!ptn.rawType().equals(HttpServerClassNames.httpServerResponseMapper)) {
            return null;
        }
        if (ptn.typeArguments().getFirst() instanceof ParameterizedTypeName entity && entity.rawType().equals(HttpServerClassNames.httpResponseEntity)) {
            var mapperTypeMirror = (DeclaredType) typeMirror;
            var entityTypeMirror = (DeclaredType) mapperTypeMirror.getTypeArguments().getFirst();
            var responseMapperElement = this.elements.getTypeElement(HttpServerClassNames.httpServerResponseMapper.canonicalName());
            var responseEntityMapperElement = this.elements.getTypeElement(HttpServerClassNames.httpServerResponseEntityMapper.canonicalName());
            var responseType = entityTypeMirror.getTypeArguments().getFirst();
            var tags = new ArrayList<String>();
            tags.add(tag);
            return () -> new ExtensionResult.CodeBlockResult(
                responseEntityMapperElement,
                dependencies -> CodeBlock.of("new $T<>($L)", HttpServerClassNames.httpServerResponseEntityMapper, dependencies),
                mapperTypeMirror,
                tag,
                List.of(this.types.getDeclaredType(responseMapperElement, responseType)),
                tags
            );

        }
        return null;
    }
}

File: http/http-server-symbol-processor/src/main/kotlin/ru/tinkoff/kora/http/server/symbol/procesor/extension/HttpServerRequestMapperKoraExtension.kt
package ru.tinkoff.kora.http.server.symbol.procesor.extension

import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSType
import com.google.devtools.ksp.symbol.Variance
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.ParameterizedTypeName
import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.kora.http.server.symbol.procesor.HttpServerClassNames
import ru.tinkoff.kora.kora.app.ksp.KoraAppUtils.findSinglePublicConstructor
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionResult
import ru.tinkoff.kora.kora.app.ksp.extension.KoraExtension
import ru.tinkoff.kora.ksp.common.KspCommonUtils.getClassDeclarationByName


class HttpServerRequestMapperKoraExtension : KoraExtension {
    override fun getDependencyGenerator(resolver: Resolver, type: KSType, tag: String?): (() -> ExtensionResult)? {
        val typeName = type.toTypeName()
        if (typeName !is ParameterizedTypeName) {
            return null
        }
        if (typeName.rawType != HttpServerClassNames.httpServerResponseMapper) {
            return null
        }
        val rsTypeName = typeName.typeArguments.first()
        val rsType = type.arguments.first().type!!.resolve()
        if (rsTypeName is ParameterizedTypeName && rsTypeName.rawType == HttpServerClassNames.httpResponseEntity) {
            val responseMapperElement = resolver.getClassDeclarationByName(HttpServerClassNames.httpServerResponseMapper)!!
            val responseEntityMapperElement = resolver.getClassDeclarationByName(HttpServerClassNames.httpServerResponseEntityMapper)!!
            val responseType = rsType.arguments.first().type!!.resolve().makeNotNullable()
            val tags = ArrayList<String?>()
            tags.add(tag)
            return {
                val delegateType = responseMapperElement.asType(listOf(resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(responseType), Variance.INVARIANT)))
                ExtensionResult.CodeBlockResult(
                    (responseEntityMapperElement.primaryConstructor ?: responseEntityMapperElement.findSinglePublicConstructor()),
                    { dependencies -> CodeBlock.of("%T(%L)", HttpServerClassNames.httpServerResponseEntityMapper, dependencies) },
                    type,
                    tag,
                    listOf(delegateType),
                    tags
                )
            }
        }
        return null
    }
}
